### PR TITLE
Make GetCourseOptionFromCodes more lenient about site_code

### PR DESCRIPTION
**Pull request by benswannack**
## Context

API integrations are not necessarily aware of Apply/Publish site codes, and so cannot provide a site code when making/changing an offer.

## Changes proposed in this pull request

This allows an API user to leave out a site code when changing an offer. If a single course option can be determined without the site, then good. Otherwise let the user know it is ambiguous.

## Guidance to review
Have I missed anything out?
I have also slightly tweaked the way we do the normal course_option finding, but I think the intention is the same.

## Link to Trello card
https://trello.com/c/d6UFJC1v/4225-relaxing-validations-for-site-codes-in-the-api

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/queries/get_course_option_from_codes.rb [hunk 1]
```diff
@@ -66,4 +57,26 @@ def initialize(
...
+    course_option_attrs = { study_mode: record.study_mode }
+
+    course_option_attrs[:site] = record.site if record.site
+
+    possible_course_options = record.course.course_options.where(course_option_attrs)
+
+    if possible_course_options.count == 1
+      record.course_option ||= possible_course_options.first
+    else
+      record.errors.add(attr, build_course_option_error_message(record, possible_course_options.count))
```
Comment (benswannack) [inline app/queries/get_course_option_from_codes.rb:71]: We now handle the "ambiguous" and "not found" cases together, which has slightly changed the behaviour when providing a site. Previously we were doing a find_by, which does not care if the found record is the only one. Now we ensure the found option is unique. This shouldn't have any impact unless we have some bad data somewhere

File: spec/queries/get_course_option_from_codes_spec.rb [hunk 2]
```diff
@@ -16,13 +16,42 @@
...
+      end
+    end
+
+    context 'when the site code is blank and ambiguous' do
+      let!(:another_course_option)  { create(:course_option, course: course_option.course) }
+
+      it 'is not valid' do
+        service.site_code = nil
+        expect(service).to be_invalid
+        expected_message = "cannot find unique #{course_option.course.study_mode} option for course #{course_option.course.code}"
```
Comment (benswannack) [inline spec/queries/get_course_option_from_codes_spec.rb:51]: Is this message clear enough? Do we want to include the word "ambiguous" somewhere?

File: app/queries/get_course_option_from_codes.rb [hunk 3]
```diff
@@ -66,4 +57,26 @@ def initialize(
...
+    end
+  end
+
+  def self.build_course_option_error_message(record, match_count)
+    error_message = match_count.zero? ? 'cannot find' : 'cannot find unique'
+
+    error_message << " #{record.study_mode} option"
```
Comment (cwrw) [inline app/queries/get_course_option_from_codes.rb:78]: I realised that we will never see this error due to the controller: https://github.com/DFE-Digital/apply-for-teacher-training/blob/main/app/controllers/vendor_api/decisions_controller.rb#L10 which just returns a generic error if it is missing. I reckon we should amend this to return the expected errors

**Review by cwrw — COMMENTED** [review]
looks good! some minor comments around messages and amending our validation surfacing

File: spec/queries/get_course_option_from_codes_spec.rb (hunk 2 above)
Comment (benswannack) [reply spec/queries/get_course_option_from_codes_spec.rb]: It's now "cannot find any full_time options for _" and "found multiple full_time options for _", which I think is clear enough

**Review by cwrw — APPROVED** [review]

**Pull request merged**